bundles/org.openhab.binding.siemensrds/src/main/java/org/openhab/binding/siemensrds/internal/RdsHandlerFactory.java

```diff
@@ -0,0 +1,125 @@
+ 
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections
+            .unmodifiableSet(new HashSet<>(Arrays.asList(THING_TYPE_CLOUD, 
+                                                         THING_TYPE_RDS)));
+
+    private final Map<ThingUID, ServiceRegistration<?>> discos = new HashMap<>();
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if ((thingTypeUID.equals(THING_TYPE_CLOUD)) && (thing instanceof Bridge)) {
+            RdsCloudHandler handler = new RdsCloudHandler((Bridge) thing);
+            createDiscoveryService(handler);
+            return handler;
+        }
+        
+        if (thingTypeUID.equals(THING_TYPE_RDS)) 
+            return new RdsHandler(thing);
+        
+        return null;
+    }
+
+    @Override
+    protected synchronized void removeHandler(ThingHandler handler) {
+        if (handler instanceof RdsCloudHandler) {
+            destroyDiscoveryService((RdsCloudHandler) handler);
+        }
+    }
+
+    /*
+     * create a discovery service so that a newly created cloud account 
+     * will find the things that it supports
+     */
+    private synchronized void createDiscoveryService(RdsCloudHandler handler) {
+        // create a new discovery service
+        RdsDiscoveryService ds = new RdsDiscoveryService(handler);
+        
+        // register the discovery service with the OpenHAB framework
+        ServiceRegistration<?> serviceReg = 
+            bundleContext.registerService(DiscoveryService.class.getName(), ds, 
+                    new Hashtable<String, Object>());
+        
+        /* 
+         * store service registration in a list 
+         * so we can destroy it when the respective hub is destroyed
+         */
+        discos.put(handler.getThing().getUID(), serviceReg);
+        
+        // finally activate the discovery service
+        ds.activate();
+    }
+    
+    /*
+     * destroy the discovery service
+     */
+    private synchronized void destroyDiscoveryService(RdsCloudHandler handler) {
+        // fetch the respective thing's service registration from our list
+        ServiceRegistration<?> serviceReg = 
+                discos.remove(handler.getThing().getUID());
+        
+        if (serviceReg != null) {
+            // retrieve the respective discovery service from the OpenHAB framework
+            RdsDiscoveryService disco = (RdsDiscoveryService) bundleContext
+                    .getService(serviceReg.getReference());
+
+            // deactivate the service
+            if (disco != null) 
+                disco.deactivate();
+
+            // and unregister the service
+            serviceReg.unregister();
```

Comment:
It actually first needs to be unregistered and then `deactivate` can be called.

Reply:
Ok.